packages/cells/src/widget.ts

```diff
@@ -955,13 +968,30 @@ export class CodeCell extends Cell<ICodeCellModel> {
   protected onStateChanged(model: ICellModel, args: IChangedArgs<any>): void {
     switch (args.name) {
       case 'executionCount':
+        this._executedCode = model.value.text.trim();
+        this._isDirty = false;
+        this.removeClass(DIRTY_CLASS);
         this.setPrompt(`${(model as ICodeCellModel).executionCount || ''}`);
         break;
       default:
         break;
     }
   }
 
+  /**
+   * Handle changes in the cell content.
+   */
+  protected onContentChanged(model: ICellModel): void {
+    if (
+      !this._isDirty &&
+      this._executedCode !== null &&
+      this._executedCode != model.value.text.trim()
+    ) {
+      this._isDirty = true;
+      this.addClass(DIRTY_CLASS);
+    }
```

Comment:
Can we make it a round-trip so that the status is cleared if the user accidentally enters a character but then reverts the change? For example:

```suggestion
    } else if (
      this._isDirty &&
      this._executedCode == model.value.text.trim()
    ) {
      this._isDirty = false;
      this.removeClass(DIRTY_CLASS);
    }
```

Reply:
That does seem like a nice UX to have. For example, I've always appreciated how smart git (and transitively most of the IDEs that integrate with it) is about recognizing when a user has done a "manual" revert.

Again though, my concern is that this potentially very large check is occurring on every user input in a cell. It's not likely to be a problem in 95% of use cases, but it's pretty easy to imagine things going wrong in the case of very large cell input. One not-particularly-nice usage pattern that I see a lot of (especially in notebooks made by recently minted quants) is the copy and paste dump of the entire contents of a CSV, which easily top `10^6` chars. I'm not saying that the "large literal dump" is good practice, this kind of usage already causes some severe performance issues. But I am concerned about performance in general, and also about making a bad situation worse, especially for users who are new to jupyter and/or python itself.

Reply:
I think that it is extremely cheap in comparison to rendering `10^6` of characters on screen. If you do as much as enter a new line, the CodeMirror will have to reflow literally millions of nodes. I would not worry about this check too much.

Reply:
You probably have a point about the relative cost of rendering. I do still feel like I need a little convincing that this is GRAS w.r.t. performance, as I *really* hate having to deal with sluggish keyboard input in code tools. I dunno. Maybe the daily trauma of having to code in a remote VM for work is getting to me.

Also, relevant to nothing, that's a slight exaggeration of the number of nodes/reflows; if they at least manage to get the csv into one string then it's only two nodes (a `<span>` nested in a `<pre>`) per line. Typically codemirror will only create a new node on the same line if/when it gets to a point where the syntax highlighting color need to change. But yes still terrible

Reply:
@telamonian what would you think of implementing what @krassowski is suggesting here, but debouncing the call?

Reply:
This one seems not resolved. 

@telamonian yes, a milion nodes is an overestimate, very much depends on whether it is syntax highlighted or wrapped into a string. Though is this really a use case that we want to support? From my perspective this an abuse of notebooks, and I would not want to have an annoying feature (orange colouring of cell that has not changed) because the system was designed with such behaviour in mind. This is different than having a long notebook or large output - those are definitely valid use cases (which we sometimes fail to support well).